python/uv/private/lock.bzl

```diff
@@ -15,109 +15,242 @@
-}) if BZLMOD_ENABLED else ["@platforms//:incompatible"]
+_REQUIREMENTS_TARGET_COMPATIBLE_WITH = [] if BZLMOD_ENABLED else ["@platforms//:incompatible"]
 
-def lock(*, name, srcs, out, upgrade = False, universal = True, args = [], **kwargs):
+_LockRunInfo = provider(
+    doc = "Information about source tree for Sphinx to build.",
+    fields = {
+        "args": "",
+        "cmd": "",
+        "cmd_file": "",
+        "srcs": "",
+    },
+)
+
+def _lock_impl(ctx):
+    args = ctx.actions.args()
+    if ctx.files.maybe_out:
+        args.add_all([
+            "--src-out",
+            ctx.files.maybe_out[0].path,
+        ])
+    args.add("--output-file", ctx.outputs.out)
+    args.add_all(ctx.files.srcs)
+    args.add_all(["--custom-compile-command", "bazel run //{}:{}.update".format(
+        ctx.label.package,
+        ctx.label.name,
+    )])
+    args.add_all([
+        "--no-python-downloads",
+        "--no-cache",
+    ])
+    args.add_all(ctx.attr.args)
+
+    # We use a manual param file so that we can forward it to the debug executable rule
+    param_file = ctx.actions.declare_file(ctx.label.name + ".params")
+    ctx.actions.write(
+        output = param_file,
+        content = args,
+    )
+
+    run_args = [param_file.path]
+    args = ctx.actions.args()
+    args.add_all(run_args)
+
+    cmd = ctx.executable._cmd
+
+    srcs = ctx.files.srcs + ctx.files.maybe_out + [param_file]
+    ctx.actions.run(
+        executable = cmd,
+        mnemonic = "RulesPythonUvPipCompile",
+        inputs = srcs,
+        outputs = [ctx.outputs.out],
+        arguments = [args],
+        tools = [cmd],
+        progress_message = "Locking requirements using uv",
+        env = ctx.attr.env,
+    )
+
+    return [
+        DefaultInfo(
+            files = depset([ctx.outputs.out]),
+        ),
+        _LockRunInfo(
+            cmd = ctx.attr._cmd,
+            cmd_file = cmd,
+            args = param_file,
+            srcs = srcs,
+        ),
+    ]
+
+_lock = rule(
+    implementation = _lock_impl,
+    doc = """\
+""",
+    attrs = {
+        "args": attr.string_list(),
+        "env": attr.string_dict(),
+        "maybe_out": attr.label(allow_single_file = True),
+        "out": attr.output(mandatory = True),
+        "run_args": attr.string_list(),
+        "srcs": attr.label_list(mandatory = True, allow_files = True),
+        "_cmd": attr.label(
+            default = "//python/uv/private:pip_compile",
+            executable = True,
+            cfg = "target",
+        ),
+    },
+)
+
+def _run_lock_impl(ctx):
+    run_info = ctx.attr.lock[_LockRunInfo]
+    params = ctx.actions.declare_file(ctx.label.name + ".params.txt")
+    ctx.actions.symlink(
+        output = params,
+        target_file = run_info.args,
+    )
+
+    executable = ctx.actions.declare_file(ctx.label.name + ".exe")
+    ctx.actions.write(
+        output = executable,
+        content = "#!/usr/bin/env bash\nset -eux\nexec {} {} \"$@\"".format(
```

Comment:
This kind of works only on bash and windows requires extra setting. Maybe just using the 2 toolchains and running the python file separately is better?

I think the problem is that we need a cross-platform way to copy the file contents from some location before we run `uv` itself. Hence, either we need to write `powershell` to do that or `bash`. Write now we have python and we still need `powershell` or `bash`. Maybe the best way would be to just use `bash` or `powershell` from the start?